### PR TITLE
Move Tower to 2018 Edition

### PR DESCRIPTION
**Pull request by davidbarsky**
Ran tests + fmt + cargo fix --edition. Everything should be on the 2018 edition now.

### CODE REVIEW

File: tower-balance/examples/demo.rs [hunk 1]
```diff
@@ -1,18 +1,12 @@
...
-extern crate tokio;
-extern crate tower;
-extern crate tower_balance;
-extern crate tower_buffer;
-extern crate tower_discover;
-extern crate tower_limit;
-extern crate tower_service;
-extern crate tower_util;
+use env_logger;
+
```
Comment (LucioFranco) [inline tower-balance/examples/demo.rs]: should be able to remove these no?

File: tower-balance/examples/demo.rs [hunk 2]
```diff
@@ -125,7 +119,7 @@ fn main() {
     rt.shutdown_on_idle().wait().unwrap();
 }
 
-type Error = Box<::std::error::Error + Send + Sync>;
+type Error = Box<dyn ::std::error::Error + Send + Sync>;
```
Comment (LucioFranco) [inline tower-balance/examples/demo.rs]: I think here we can `Box<dyn std::error::Error + Send + Sync>`

File: tower-balance/src/lib.rs [hunk 3]
```diff
@@ -1,19 +1,11 @@
...
-extern crate futures;
-#[macro_use]
-extern crate log;
-extern crate indexmap;
-extern crate rand;
-extern crate tokio_timer;
-extern crate tower_discover;
-extern crate tower_service;
-extern crate tower_util;
+use rand;
```
Comment (LucioFranco) [inline tower-balance/src/lib.rs]: same here i dont think we need this use?

File: tower-balance/src/load/peak_ewma.rs [hunk 4]
```diff
@@ -307,8 +308,8 @@ fn nanos(d: Duration) -> f64 {
 
 #[cfg(test)]
 mod tests {
-    extern crate tokio_executor;
-    extern crate tokio_timer;
+    use tokio_executor;
```
Comment (LucioFranco) [inline tower-balance/src/load/peak_ewma.rs]: same here?

File: tower-buffer/src/error.rs [hunk 5]
```diff
@@ -22,7 +22,7 @@ pub struct SpawnError {
 }
 
 /// Errors produced by `Buffer`.
-pub(crate) type Error = Box<::std::error::Error + Send + Sync>;
+pub(crate) type Error = Box<dyn ::std::error::Error + Send + Sync>;
```
Comment (LucioFranco) [inline tower-buffer/src/error.rs]: again here we can drop the `::`

File: tower-buffer/tests/buffer.rs [hunk 6]
```diff
@@ -1,8 +1,4 @@
-extern crate futures;
-extern crate tokio_executor;
-extern crate tower_buffer;
-extern crate tower_mock;
-extern crate tower_service;
+use tower_mock;
```
Comment (LucioFranco) [inline tower-buffer/tests/buffer.rs]: same here

File: tower-filter/tests/filter.rs [hunk 7]
```diff
@@ -1,7 +1,4 @@
-extern crate futures;
-extern crate tower_filter;
-extern crate tower_mock;
-extern crate tower_service;
+use tower_mock;
```
Comment (LucioFranco) [inline tower-filter/tests/filter.rs]: same here

File: tower-limit/tests/concurrency.rs [hunk 8]
```diff
@@ -1,8 +1,6 @@
-extern crate futures;
-extern crate tokio_mock_task;
-extern crate tower_limit;
-extern crate tower_mock;
-extern crate tower_service;
+use futures;
```
Comment (LucioFranco) [inline tower-limit/tests/concurrency.rs]: same here

File: tower-limit/tests/rate.rs [hunk 9]
```diff
@@ -1,9 +1,7 @@
-extern crate futures;
-extern crate tokio;
-extern crate tokio_timer;
-extern crate tower_limit;
-extern crate tower_mock;
-extern crate tower_service;
+use tokio;
```
Comment (LucioFranco) [inline tower-limit/tests/rate.rs]: and here

File: tower-load-shed/tests/load-shed.rs [hunk 10]
```diff
@@ -1,7 +1,5 @@
-extern crate futures;
-extern crate tower_load_shed;
-extern crate tower_mock;
-extern crate tower_service;
+use tower_load_shed;
```
Comment (LucioFranco) [inline tower-load-shed/tests/load-shed.rs]: here too!

File: tower-mock/tests/mock.rs [hunk 11]
```diff
@@ -1,6 +1,4 @@
-extern crate futures;
-extern crate tower_mock;
-extern crate tower_service;
+use tower_mock;
```
Comment (LucioFranco) [inline tower-mock/tests/mock.rs]: here aswell :D

File: tower-reconnect/src/lib.rs [hunk 12]
```diff
@@ -23,7 +18,7 @@ where
     target: Target,
 }
 
-type Error = Box<::std::error::Error + Send + Sync>;
+type Error = Box<dyn ::std::error::Error + Send + Sync>;
```
Comment (LucioFranco) [inline tower-reconnect/src/lib.rs]: can drop the `::`

File: tower-retry/src/budget.rs [hunk 13]
```diff
@@ -212,7 +212,7 @@ impl Bucket {
 
 #[cfg(test)]
 mod tests {
-    extern crate tokio_executor;
+    use tokio_executor;
```
Comment (LucioFranco) [inline tower-retry/src/budget.rs]: here too

File: tower-retry/tests/retry.rs [hunk 14]
```diff
@@ -1,7 +1,5 @@
-extern crate futures;
-extern crate tower_mock;
-extern crate tower_retry;
-extern crate tower_service;
+use tower_mock;
```
Comment (LucioFranco) [inline tower-retry/tests/retry.rs]: here tooo

File: tower-util/tests/call_all.rs [hunk 15]
```diff
@@ -1,9 +1,5 @@
-extern crate futures;
-extern crate tokio_mock_task;
-extern crate tower;
-extern crate tower_mock;
-extern crate tower_service;
-extern crate tower_util;
+use futures;
```
Comment (LucioFranco) [inline tower-util/tests/call_all.rs]: here

File: tower/examples/channel_service.rs [hunk 16]
```diff
@@ -9,16 +9,10 @@
...
-extern crate futures_cpupool;
-extern crate tokio_timer;
-extern crate tower;
-extern crate tower_service;
-
 #[macro_use]
 extern crate log;
 
-extern crate env_logger;
+use env_logger;
```
Comment (LucioFranco) [inline tower/examples/channel_service.rs]: here too

File: tower/examples/client.rs [hunk 17]
```diff
@@ -1,12 +1,5 @@
-extern crate futures;
-extern crate hyper;
-extern crate tower;
-extern crate tower_buffer;
-extern crate tower_hyper;
-extern crate tower_limit;
-extern crate tower_reconnect;
-extern crate tower_retry;
-extern crate tower_service;
+use futures;
```
Comment (LucioFranco) [inline tower/examples/client.rs]: same

File: tower/examples/server.rs [hunk 18]
```diff
@@ -1,10 +1,4 @@
-extern crate futures;
-extern crate hyper;
-extern crate tokio_tcp;
-extern crate tower;
-extern crate tower_hyper;
-extern crate tower_limit;
-extern crate tower_service;
+use hyper;
```
Comment (LucioFranco) [inline tower/examples/server.rs]: same

File: tower/src/util.rs [hunk 19]
```diff
@@ -15,7 +15,7 @@ use tower_service::Service;
 
 impl<T: ?Sized, Request> ServiceExt<Request> for T where T: Service<Request> {}
 
-type Error = Box<::std::error::Error + Send + Sync>;
+type Error = Box<dyn ::std::error::Error + Send + Sync>;
```
Comment (LucioFranco) [inline tower/src/util.rs]: can drop the `::`

File: tower/tests/builder.rs [hunk 20]
```diff
@@ -1,12 +1,4 @@
-extern crate futures;
-extern crate tokio;
-extern crate tower;
-extern crate tower_buffer;
-extern crate tower_limit;
-extern crate tower_reconnect;
-extern crate tower_retry;
-extern crate tower_service;
-extern crate void;
+use tokio;
```
Comment (LucioFranco) [inline tower/tests/builder.rs]: same here

File: tower-balance/examples/demo.rs (hunk 2 above)
Comment (davidbarsky) [reply tower-balance/examples/demo.rs]: Good point, fixed.

File: tower-balance/examples/demo.rs (hunk 1 above)
Comment (davidbarsky) [reply tower-balance/examples/demo.rs]: Yep, removed the places/merged imports.

File: tower-balance/src/load/peak_ewma.rs (hunk 4 above)
Comment (davidbarsky) [reply tower-balance/src/load/peak_ewma.rs]: not to the best of my knowledge—it's certainly nice, but this is a large diff as-is and i'd be happy to address in subsequent PRs.

File: tower-buffer/src/error.rs (hunk 5 above)
Comment (davidbarsky) [reply tower-buffer/src/error.rs]: done.

File: tower-balance/src/lib.rs (hunk 3 above)
Comment (davidbarsky) [reply tower-balance/src/lib.rs]: done.

File: tower-balance/examples/demo.rs (hunk 1 above)
Comment (davidbarsky) [reply tower-balance/examples/demo.rs]: done.

**Comment by olix0r** [comment]
Is the `<'_>` idiom enforced by rustfmt or is this something that will diverge over time?

**Comment by davidbarsky** [comment]
@LucioFranco I think I addressed all the weird spaces introduce by cargo fix.

**Comment by davidbarsky** [comment]
> Is the <'_> idiom enforced by rustfmt or is this something that will diverge over time?

@olix0r Not sure what what you mean? It's a 2018 edition thing: https://doc.rust-lang.org/edition-guide/rust-2018/ownership-and-lifetimes/the-anonymous-lifetime.html

**Comment by LucioFranco** [comment]
@olix0r I believe it's only enforced if you explicitly enable the 2018 idioms lint.

@davidbarsky are you sure you can't remove the `use hyper`? I thought you can reference a crate directly now like `hyper::Body`?

**Comment by olix0r** [comment]
@davidbarsky I'm just mean to confirm that we have some automated way to enforce the new idiom. As long as PRs that use the old implicit form will fail to pass now, I'm happy.

**Comment by davidbarsky** [comment]
I think Lucio’s right. It’s a Clippy lint that’ll need to be setup.


________________________________

**Comment by LucioFranco** [comment]
@davidbarsky its not a clippy lint, its a regular rust one, its just not enabled by default.

**Comment by davidbarsky** [comment]
My bad, I’ll enable that lint. The use-hyper is removable.

File: tower-service/Cargo.toml [hunk 21]
```diff
@@ -17,6 +17,7 @@ description = """
 Trait representing an asynchronous, request / response based, client or server.
 """
 categories = ["asynchronous", "network-programming"]
+edition = "2018"
```
Comment (carllerche) [inline tower-service/Cargo.toml]: `tower-service` cannot be switched to 2018. Doing so will prevent crates that are unable to update their minimum rust version to implement `Service`.

Comment (davidbarsky) [reply tower-service/Cargo.toml]: I can bump the version number to 0.3, signifying a breaking change, or bring it down to the 2015 edition. Which would you prefer?

Comment (hawkw) [reply tower-service/Cargo.toml]: IMHO this is not worth a breaking change right now.

Comment (davidbarsky) [reply tower-service/Cargo.toml]: Understood, I'll bump it down to 2015.

File: tower/tests/builder.rs (hunk 20 above)
Comment (davidbarsky) [reply tower/tests/builder.rs]: fixed.

**Comment by carllerche** [comment]
Sorry... I conflicted :(

**Comment by carllerche** [comment]
@olix0r @davidbarsky @LucioFranco Do we need `<'_>`? It feels like unnecessary noise. What does it gain?

File: tower-balance/src/choose/mod.rs [hunk 22]
```diff
@@ -13,12 +12,12 @@ pub trait Choose<K, N> {
...
     /// `replicas` cannot be empty, so this function must always return a valid index on
     /// [0, replicas.len()-1].
-    fn choose(&mut self, replicas: Replicas<K, N>) -> usize;
+    fn choose(&mut self, replicas: Replicas<'_, K, N>) -> usize;
 }
 
 /// Creates a `Replicas` if there are two or more services.
 ///
-pub(crate) fn replicas<K, S>(inner: &IndexMap<K, S>) -> Result<Replicas<K, S>, TooFew> {
+pub(crate) fn replicas<K, S>(inner: &IndexMap<K, S>) -> Result<Replicas<'_, K, S>, TooFew> {
```
Comment (carllerche) [inline tower-balance/src/choose/mod.rs]: It feels wonky to have this here, but not on `& '_ IndexMap<...>`

File: tower-buffer/tests/buffer.rs [hunk 23]
```diff
@@ -1,16 +1,11 @@
...
-extern crate tower_buffer;
-extern crate tower_mock;
-extern crate tower_service;
-
 use futures::prelude::*;
 use tokio_executor::{SpawnError, TypedExecutor};
-use tower_buffer::*;
-use tower_service::*;
+use tower::{
+    buffer::{error::ServiceError as BufferServiceError, Buffer},
```
Comment (carllerche) [inline tower-buffer/tests/buffer.rs]: Why is this being renamed?

Comment (davidbarsky) [reply tower-buffer/tests/buffer.rs]: We had an `error::ServiceError` being used in these tests, and it was ambiguous as to where it was being sourced from. `BufferServiceError` makes it clearer out of context.

**Comment by davidbarsky** [comment]
@carllerche Immediately, the `'_` lifetime prevents this error:

```
error: hidden lifetime parameters in types are deprecated
  --> tower-balance/src/choose/mod.rs:15:36
   |
15 |     fn choose(&mut self, replicas: Replicas<K, N>) -> usize;
   |                                    ^^^^^^^^^^^^^^ help: indicate the anonymous lifetime: `Replicas<'_, K, N>`

error: hidden lifetime parameters in types are deprecated
  --> tower-balance/src/choose/mod.rs:20:64
   |
20 | pub(crate) fn replicas<K, S>(inner: &IndexMap<K, S>) -> Result<Replicas<K, S>, TooFew> {
   |                                                                ^^^^^^^^^^^^^^ help: indicate the anonymous lifetime: `Replicas<'_, K, S>`

error: aborting due to 2 previous errors
```

I think we might be able to refactor away from this, but in the interest of minimizing changes to _only_ the 2018 edition move, I'd prefer to hold off on larger refactors.

File: tower-balance/src/choose/mod.rs (hunk 22 above)
Comment (davidbarsky) [reply tower-balance/src/choose/mod.rs]: In all other places, rustc is able to infer the lifetimes. I can add more explicit lifetime annotations like:

```rust
pub(crate) fn replicas<K, S>(inner: &'_ IndexMap<K, S>) -> Result<Replicas<'_, K, S>, TooFew>
```

...and a lifetime parameter to the `Choose` trait as such:

```rust
pub trait Choose<K, N> {
    /// Returns the index of a replica to be used next.
    ///
    /// `replicas` cannot be empty, so this function must always return a valid index on
    /// [0, replicas.len()-1].
    fn choose(&mut self, replicas: Replicas<'_, K, N>) -> usize;
}
```

...but we don't need those explicit annotations in those locations.

**Comment by davidbarsky** [comment]
Oh, and before I forget, this is the anonymous lifetime feature in 2018: https://doc.rust-lang.org/edition-guide/rust-2018/ownership-and-lifetimes/the-anonymous-lifetime.html

**Comment by LucioFranco** [comment]
@carllerche Im partially okay with the lifetime thing, as it does make it a bit more clear that there is something being borrowed there. Though im indifferent to it, it might be nice for newer people 🤷‍♂️

**Comment by seanmonstar** [comment]
I'm also personally not a fan of the anonymous lifetime noise, it doesn't provide anything but just adds noise. Yay!

File: tower-buffer/tests/buffer.rs (hunk 23 above)
Comment (carllerche) [reply tower-buffer/tests/buffer.rs]: Can the ambiguity be removed using the module instead? I'm not a huge fan of renaming types as that tends to be more confusing IMO (the type is referenced using a different name).

Comment (davidbarsky) [reply tower-buffer/tests/buffer.rs]: Yep, we can do that!

Comment (davidbarsky) [reply tower-buffer/tests/buffer.rs]: Done.

**Comment by davidbarsky** [comment]
Sign-off would be useful before I try to rebase atop of master. I _think_ I addressed most outstanding issues.

**Comment by seanmonstar** [comment]
What if we hold off on the anonymous lifetime change? It's not yet a warning from rustc, and many feel it's just distracting.

**Comment by carllerche** [comment]
I'm 👍 w/ the PR without the `'_` change. As @seanmonstar suggested, we can punt on it and figure out wtf to do later.

**Comment by davidbarsky** [comment]
@carllerche Unfortunately, the whole thing fails to compile without that single `'_` lifetime. I could either keep that single annotation or re-annotate everything there.

File: tower-balance/src/lib.rs [hunk 24]
```diff
@@ -1,4 +1,5 @@
 #![deny(rust_2018_idioms)]
+#![allow(elided_lifetimes_in_paths)]
```
Comment (seanmonstar) [inline tower-balance/src/lib.rs]: Oh, it's because of the above `deny`. Do we need it?

File: tower-balance/src/error.rs [hunk 25]
```diff
@@ -6,7 +6,7 @@ pub(crate) type Error = Box<dyn std::error::Error + Send + Sync>;
 pub struct Balance(pub(crate) Error);
 
 impl fmt::Display for Balance {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```
Comment (carllerche) [inline tower-balance/src/error.rs]: Can this lifetime be removed?

File: tower-balance/src/lib.rs (hunk 24 above)
Comment (carllerche) [reply tower-balance/src/lib.rs]: It seems good to enforce style we care about?

Comment (seanmonstar) [reply tower-balance/src/lib.rs]: Eh, I don't care either way.

File: tower-balance/src/error.rs (hunk 25 above)
Comment (davidbarsky) [reply tower-balance/src/error.rs]: remove it and friends.

**Pull request merged**